pkg/virt-controller/watch/clone/clone.go

```diff
@@ -0,0 +1,365 @@
+		_, exists, err := ctrl.vmInformer.GetStore().GetByKey(getKey(targetVMInfo.Name, vmClone.Namespace))
+		if !exists {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("target VM %s is not created yet for clone %s", targetVMInfo.Name, vmClone.Name))
+		} else if err != nil {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("error getting VM %s from cache for clone %s: %v", *vmClone.Status.SnapshotName, targetVMInfo.Name, err))
+		}
+
+		ctrl.logAndRecord(vmClone, TargetVMCreated, fmt.Sprintf("created target VM %s for clone %s", targetVMInfo.Name, vmClone.Name))
+		syncInfo.targetVMCreated = true
+
+		// Clean up snapshot & restore
+		err = ctrl.client.VirtualMachineSnapshot(vmClone.Namespace).Delete(context.Background(), *vmClone.Status.SnapshotName, v1.DeleteOptions{})
+		if !errors.IsNotFound(err) && err != nil {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("cannot clean up snapshot %s for clone %s", *vmClone.Status.SnapshotName, vmClone.Name))
+		}
+
+		err = ctrl.client.VirtualMachineRestore(vmClone.Namespace).Delete(context.Background(), *vmClone.Status.RestoreName, v1.DeleteOptions{})
+		if !errors.IsNotFound(err) && err != nil {
+			return addErrorToSyncInfo(syncInfo, fmt.Errorf("cannot clean up restore %s for clone %s", *vmClone.Status.RestoreName, vmClone.Name))
+		}
+
+	default:
+		log.Log.Object(vmClone).Infof("clone %s is in phase %s - nothing to do", vmClone.Name, string(vmClone.Status.Phase))
+	}
+
+	return syncInfo
+}
+
+func (ctrl *VMCloneController) updateStatus(origClone *clonev1alpha1.VirtualMachineClone, syncInfo syncInfo) error {
+	vmClone := origClone.DeepCopy()
+
+	var phaseChanged bool
+	assignPhase := func(phase clonev1alpha1.VirtualMachineClonePhase) {
+		vmClone.Status.Phase = phase
+		phaseChanged = true
+	}
+
+	if syncInfo.isCloneFailing {
+		ctrl.logAndRecord(vmClone, syncInfo.failEvent, syncInfo.failReason)
+		assignPhase(clonev1alpha1.Failed)
+		updateCloneConditions(vmClone,
+			newFailureCondition(corev1.ConditionTrue, fmt.Sprintf("Failed: %s", syncInfo.failReason)),
+			newProgressingCondition(corev1.ConditionFalse, "Failed"),
+			newReadyCondition(corev1.ConditionFalse, "Failed"),
+		)
+	}
+
+	updateCloneConditions(vmClone,
+		newFailureCondition(corev1.ConditionFalse, "Still processing"),
+		newProgressingCondition(corev1.ConditionTrue, "Still processing"),
+		newReadyCondition(corev1.ConditionFalse, "Still processing"),
+	)
+
+	if isInPhase(vmClone, clonev1alpha1.PhaseUnset) {
+		assignPhase(clonev1alpha1.SnapshotInProgress)
+	}
+	if isInPhase(vmClone, clonev1alpha1.SnapshotInProgress) {
+		if snapshotName := syncInfo.snapshotName; snapshotName != "" {
+			vmClone.Status.SnapshotName = pointer.String(snapshotName)
+		}
+
+		if syncInfo.snapshotReady {
+			assignPhase(clonev1alpha1.RestoreInProgress)
+		}
+	}
+	if isInPhase(vmClone, clonev1alpha1.RestoreInProgress) {
+		if restoreName := syncInfo.restoreName; restoreName != "" {
+			vmClone.Status.RestoreName = pointer.String(restoreName)
```

Comment:
Should restoreName/SnapshotName be removed from the status once they are deleted?

Reply:
I'm not sure, since this info may be valuable for debugging?
But on the other hand, we do have this info in our logs. So yeah, I guess they can be removed.